crates/paralegal-policy/src/context.rs

```diff
@@ -40,10 +40,118 @@ pub enum EdgeType {
     DataAndControl,
 }
 
+/// Data type representing nodes in the SPDG for a particular controller.
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub struct CtrlNode<'a> {
+    /// ControllerId of the node.
+    pub ctrl_id: &'a ControllerId,
+    /// The data of the node.
+    pub node: Node<'a>,
+}
+
+/// Enum identifying the different types of nodes in the SPDG
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub enum Node<'a> {
+    /// Corresponds to [`DataSource::Argument`].
+    CtrlArgument(&'a DataSource),
+
+    /// Corresponds to a [`CallSite`] or [`DataSource::FunctionCall`].
+    CallSite(&'a CallSite),
+
+    /// Corresponds to a [`DataSink::Argument`].
+    CallArgument(&'a DataSink),
+
+    /// Corresponds to a [`DataSink::Return`].
+    Return(&'a DataSink),
+}
+
+impl<'a> From<&'a CallSite> for Node<'a> {
+    fn from(cs: &'a CallSite) -> Self {
+        Node::CallSite(cs)
+    }
+}
+
+impl<'a> From<&'a DataSink> for Node<'a> {
+    fn from(ds: &'a DataSink) -> Self {
+        match ds {
+            DataSink::Argument { .. } => Node::CallArgument(ds),
+            DataSink::Return => Node::Return(ds),
+        }
+    }
+}
+
+impl<'a> From<&'a CallSiteOrDataSink> for Node<'a> {
+    fn from(x: &'a CallSiteOrDataSink) -> Self {
+        match x {
+            CallSiteOrDataSink::CallSite(cs) => cs.into(),
+            CallSiteOrDataSink::DataSink(ds) => ds.into(),
+        }
+    }
+}
+
+impl<'a> From<&'a DataSource> for Node<'a> {
+    fn from(ds: &'a DataSource) -> Self {
+        match ds {
+            DataSource::FunctionCall(cs) => Node::CallSite(cs),
+            DataSource::Argument(_) => Node::CtrlArgument(ds),
+        }
+    }
+}
+
+impl<'a> Node<'a> {
+    /// Transform a Node into it's corresponding CallSite - only works for CallSites and CallArguments.
+    pub fn as_call_site(&self) -> Option<&CallSite> {
```

Comment:
Done

Reply:
`&'a CallSite` means that it is tied to the lifetime `'a`. If you use just `&` it is tied to the implicit lifetime `'_`, which in this case is the lifetime of `self`. Since `self` also entails `'a` this constraint is stricter.

Perhaps as a contrived example you wouldn't be able to do this

```rs
let cs = CallSite(...);
let the_ref = Node::CallSite(&cs).as_call_site();
```

because the `Node`, constructed inline, only lives as long as the expression to the right of the `=`. So it's lifetime is short, whereas `cs` (our lifetime `'a`) is `let` bound and lives much longer. So to make this valid you'd have to do

```rs
let cs = CallSite(...);
let n = Node::CallSite(&cs);
let the_ref = n.as_call_site();
```

This is fine for simple values like these usually, but can be quite annoying when dealing with iterators, which may then have to be collected and iterated again to make sure the lifetimes work out.

I wanted to make sure I wasn't telling you nonsense so I tried it out in [the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=38b873f037fb918dab135c9a037f08ae). The condensed example I wrote up there doesn't compile but if you add the `'s` lifetime to the `as_s` method it does and you can see the error it has at first which tells you kinda the same thing I tried to explain also.

Reply:
Ah, I think I get it. Thanks for the explanation!